mm2src/mm2_metrics/src/adapt.rs

```diff
@@ -0,0 +1,381 @@
+            if let Some(recorder) = $crate::new_lib::TryRecorder::try_recorder(&$metrics){
+            let key_name = $crate::metrics::KeyName::from_const_str(&$name);
+            let key = $crate::metrics::Key::from_name(key_name);
+            let gauge = recorder.register_gauge(&key);
+            gauge.increment($value);
+        }
+    }};
+
+    // Register and increment gauge with label
+    ($metrics:expr, $name:expr, $value:expr, $($label_key:expr => $label_val:expr),+) => {{
+         if let Some(recorder) = $crate::new_lib::TryRecorder::try_recorder(&$metrics){
+            let key_name = $crate::metrics::KeyName::from_const_str(&$name);
+            let key = $crate::metrics::Key::from_parts(key_name, from_slice_to_labels(&[($($label_key, $label_val),+)]));
+            let gauge = recorder.register_gauge(&key);
+            gauge.increment($value);
+        }
+    }};
+}
+
+/// Update gauge if an MmArc is not dropped yet and metrics system is initialized already.
+#[macro_export]
+macro_rules! mm_timing_new {
+    ($metrics:expr, $name:expr, $value:expr) => {{
+            if let Some(recorder) = $crate::new_lib::TryRecorder::try_recorder(&$metrics){
+            let key_name = $crate::metrics::KeyName::from_const_str(&$name);
+            let key = $crate::metrics::Key::from_name(key_name);
+            let histo = recorder.register_histogram(&key);
+            histo.record($value);
+        }
+
+    }};
+
+    // Register and record histogram with label
+    ($metrics:expr, $name:expr, $value:expr, $($label_key:expr => $label_val:expr),+) => {{
+         if let Some(recorder) = $crate::new_lib::TryRecorder::try_recorder(&$metrics){
+            let key_name = $crate::metrics::KeyName::from_const_str(&$name);
+            let key = $crate::metrics::Key::from_parts(key_name, from_slice_to_labels(&[($($label_key, $label_val),+)]));
+            let histo = recorder.register_histogram(&key);
+            histo.record($value);
+         }
+    }};
+}
+
+#[allow(dead_code)]
+fn from_slice_to_labels(labels: &[(&'static str, &'static str)]) -> Vec<Label> { labels.into_labels() }
+
+pub struct Snapshot {
+    pub counters: HashMap<String, HashMap<Vec<String>, u64>>,
+    pub gauges: HashMap<String, HashMap<Vec<String>, u64>>,
+    pub histograms: HashMap<String, HashMap<Vec<String>, f64>>,
+}
+
+#[allow(dead_code)]
+pub(crate) struct Inner {
+    pub registry: Registry<metrics::Key, GenerationalAtomicStorage>,
+}
+
+impl Inner {
+    fn get_metrics(&self) -> Snapshot {
+        let mut counters = HashMap::new();
+        for (key, counter) in self.registry.get_counter_handles() {
+            let value = counter.get_inner().load(Ordering::Acquire);
+            let (_name, labels) = key_to_parts(&key, None);
+            let entry = counters
+                .entry(key.to_string())
+                .or_insert_with(HashMap::new)
+                .entry(labels)
+                .or_insert(value);
+            *entry = value;
```

Comment:
It seems to be possible to move this implementation into a generic function.
So I consider something like this:
```rust
let counters = self
  .registry
  .get_counter_handles()
  .into_iter()
  .map(|(key, counter)| {
    let value = counter.get_inner().load(Ordering::Acquire); // This is a specific part for counter/gauge and histogram.
    key_value_to_snapshot_entry(key, value)
  })
  .collect();
```

Reply:
basically there's much difference in moving it ...

here's the same code after moving 

```rust
 let counters = self
            .registry
            .get_counter_handles()
            .into_iter()
            .map(|(key, counter)| {
                let value = counter.get_inner().load(Ordering::Acquire); 
                let inner = key_value_to_snapshot_entry(key.clone(), value);
                (key.to_string(), inner)
            })
            .collect::<HashMap<String, HashMap<_, _>>>();
```